File: py/graph/shortest_path/FloydWarshall.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys
from typing import List

class EdgeNode:
    def __init__(self, index: int, weight: int):
        self.index, self.weight, self.next = index, weight, None

class VertexNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class FloydWarshall:
    def search(self, graphs: List[VertexNode]) -> List[List[int]]:
        result = [[sys.maxsize for _ in range(len(graphs))] for _ in range(len(graphs))]
        for i, value in enumerate(graphs):
            edge = value.next
            while edge:
                result[i][edge.index] = edge.weight
                edge = edge.next
            result[i][i] = 0

        for i in range(len(graphs)):
            for j in range(len(graphs)):
                for k in range(len(graphs)):
                    if sys.maxsize == result[j][i] or sys.maxsize == result[i][k]:
                        continue

                    pivot = result[j][i] + result[i][k]
                    if pivot < result[j][k]:
                        result[j][k] = pivot
        return result

File: java/graph/shortest_path/FloydWarshall.java
package graph.shortest_path;

import java.util.Arrays;

public class FloydWarshall {
    public int[][] search(VertexNode[] graphs) {
        int[][] result = new int[graphs.length][graphs.length];
        for (int i = 0; i < graphs.length; ++i) {
            Arrays.fill(result[i], Integer.MAX_VALUE);
            EdgeNode edge = graphs[i].next;
            while (null != edge) {
                result[i][edge.index] = edge.weight;
                edge = edge.next;
            }
            result[i][i] = 0;
        }

        for (int i = 0; i < graphs.length; ++i) {
            for (int j = 0; j < graphs.length; ++j) {
                for (int k = 0; k < graphs.length; ++k) {
                    if (Integer.MAX_VALUE == result[j][i] || Integer.MAX_VALUE == result[i][k]) {
                        continue;
                    }

                    int pivot = result[j][i] + result[i][k];
                    if (pivot < result[j][k]) {
                        result[j][k] = pivot;
                    }
                }
            }
        }
        return result;
    }

    public class EdgeNode {
        public int index = 0;
        public int weight = 0;
        public EdgeNode next = null;

        public EdgeNode(int index, int weight) {
            this.index = index;
            this.weight = weight;
        }
    }

    public class VertexNode {
        public int value = 0;
        public EdgeNode next = null;

        public VertexNode(int value) {
            this.value = value;
        }
    }
}
